ocaml/parsing/jane_syntax.ml

```diff
@@ -400,35 +400,59 @@ end = struct
   end
 end
 
-module Mode_annotation = struct
-  type t =
-    | Local
-    | Unique
-    | Once
+module Mode_expr = struct
+  type t = string Location.loc list Location.loc
+
+  let empty = Location.mknoloc []
+
+  let singleton mode loc = Location.mkloc [Location.mkloc mode loc] loc
+
+  let is_empty { txt; _ } = match txt with [] -> true | _ -> false
 
-  let equal t1 t2 =
-    match t1, t2 with
-    | Local, Local -> true
-    | Unique, Unique -> true
-    | Once, Once -> true
-    | (Local | Unique | Once), _ -> false
+  let embedded_name = Embedded_name.of_feature (Language_extension Mode) []
 
-  include Make_payload_protocol_of_stringable (struct
-    type nonrec t = t
+  let embedded_name_str = Embedded_name.to_string embedded_name
+
+  module Payload_protocol_of_mode = Make_payload_protocol_of_stringable (struct
+    type t = string
 
     let indefinite_article_and_name = "a", "mode"
 
-    let to_string = function
-      | Local -> "local"
-      | Unique -> "unique"
-      | Once -> "once"
+    let to_string s = s
 
-    let of_string = function
-      | "local" -> Some Local
-      | "unique" -> Some Unique
-      | "once" -> Some Once
-      | _ -> None
+    let of_string s = Some s
   end)
+
+  let payload_of { txt; _ } =
+    Payload_protocol_of_mode.Encode.list_as_payload txt
+
+  let of_payload ~loc payload =
+    let l = Payload_protocol_of_mode.Decode.list_from_payload ~loc payload in
+    Location.mkloc l loc
+
+  let partition_attrs attrs =
+    List.partition
+      (fun { attr_name; _ } -> attr_name.txt = embedded_name_str)
+      attrs
```

Comment:
Should `partition_attrs` return `Parsetree.attribute option * Parsetree.attribute list` instead of `Parsetree.attribute list * Parsetree.attribute list`? It appears that `of_attrs` blindly `assert`s `false` if the first attribute list isn't empty or a singleton list, and it's unclear why it would have different preconditions than `partition_attrs`.

Separately, I would much prefer a `Misc.failwith` rather than `assert false` as it lets you explain why this code is unreachable.

Reply:
Indeed, our assumption is that there should be at most one mode attribute at each place. 

The interface of `partition_attrs` is like so to avoid a roundtrip in `typecore.ml`.

But I agree clarity of interface is more important.